File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> res;
        for(int num : nums){
            if(res.empty() || res.back() < num){
                res.push_back(num);
            }
            else {
                int index = find(res, num);
                res[index] = num;
            }
        }
        return res.size();
    }
    int find(vector<int> v, int target){
        int l = 0;
        int r = v.size()-1;
        while(l <= r){
            int mid = (r+l) / 2;
            if(v[mid] == target){
                return mid;
            }
            else if(v[mid] > target){
                r = mid-1;
            }
            else {
                l = mid+1;
            }
        }
        return l;
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.dart
class Solution {
  int lengthOfLIS(List<int> nums) {
    int count = 0;
    List<int> list = [];
    for(int i=0;i<nums.length;i++){
        if(list.length == 0 || list.last < nums[i]){
            list.add(nums[i]);
        }else {
            int index = findIndex(list, nums[i]);
            list[index] = nums[i];
        }
    }
    return list.length;
  }
  int findIndex(List<int> list, int target){
    int left = 0, right = list.length-1;
    while(left <= right){
        int mid = (right+left) ~/ 2;
        if(list[mid] == target){
            return mid;
        } else if(list[mid] < target){
            left = mid+1;
        } else {
            right = mid-1;
        }
    }
    return left;
  }
}